src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,655 @@
+                    remainingChars = remainingChars.Slice(next);
+                } while (!isFinalSegment);
+
+                return originalBytesLength - remainingBytes.Length; // total number of bytes we wrote
+            }
+        }
+
+        /// <summary>
+        /// Encodes the specified <see cref="ReadOnlySequence{Char}"/> into a <see cref="byte"/> array using the specified <see cref="Encoding"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="chars"/> should be encoded.</param>
+        /// <param name="chars">The <see cref="ReadOnlySequence{Char}"/> to encode to <see langword="byte"/>s.</param>
+        /// <returns>A <see cref="byte"/> array which represents the encoded contents of <paramref name="chars"/>.</returns>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static byte[] GetBytes(this Encoding encoding, in ReadOnlySequence<char> chars)
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (chars.IsSingleSegment)
+            {
+                // If the incoming sequence is single-segment, one-shot this.
+
+                ReadOnlySpan<char> span = chars.FirstSpan;
+
+                byte[] retVal = new byte[encoding!.GetByteCount(span)];
+                int actualBytesWritten = encoding.GetBytes(span, retVal);
+
+                Debug.Assert(actualBytesWritten == retVal.Length, "Unexpected number of elements written.");
+
+                return retVal;
+            }
+            else
+            {
+                // If the incoming sequence is multi-segment, create a stateful Encoder
+                // and use it as the workhorse. On the final iteration we'll pass flush=true.
+
+                Encoder encoder = encoding!.GetEncoder();
+
+                // Maintain a list of all the segments we'll need to concat together.
+                // These will be released back to the pool at the end of the method.
+
+                List<(byte[], int)> listOfSegments = new List<(byte[], int)>(8); // guess the initial capacity
+                int totalByteCount = 0;
+
+                ReadOnlySequence<char> remainingChars = chars;
+                bool isFinalSegment;
+
+                do
+                {
+                    remainingChars.GetFirstSpan(out ReadOnlySpan<char> firstSpan, out SequencePosition next);
+                    isFinalSegment = remainingChars.IsSingleSegment;
+
+                    int byteCountThisIteration = encoder.GetByteCount(firstSpan, flush: isFinalSegment);
+                    byte[] rentedArray = ArrayPool<byte>.Shared.Rent(byteCountThisIteration);
```

Comment:
This could potentially be renting a lot of arrays?  Is there any concern about one operation draining the pool?

Reply:
Yeah, it'll rent in total a greater number of bytes than it returns. The scratch buffer needed for this operation is potentially significant and I didn't want to put pressure on the GC. Perhaps it'd make sense to use native memory for the intermediate buffers?

Reply:
Spoke with David offline. The common scenario for this is that the input sequence will be single-segment (this optimization is handled earlier in the method), or that if it is multi-segment the final object will almost always be small enough to fit into the Gen 0 heap. Of course this would need to work for long sequence as well, but that shouldn't be a common case.

Given the target scenarios I think it's ok that we're using the array pool in this fashion.

Reply:
Sounds good.

But can you just clarify what you mean by "it'll rent in total a greater number of bytes than it returns". It's not returning to the pool everything it rents from the pool?

Reply:
Oof, that's confusing phrasing conflating "returning to the pool" and "returning from the method". :)

What this means is that we might end up renting from (and returning to) the array pool a total of 2 MB of scratch buffer space, even if the object we're going to return from the method is `new byte[1024 * 1024]`. If this is a problem, there's an alternative implementation of this method that is much more conservative with how many buffers are rented from the array pool. The tradeoff is that we'll end up walking the incoming sequence more times since we'll have to recompute some things.

In non-exceptional cases, all rented arrays are returned to the array pool. If an exception occurs, the rented arrays are abandoned and will eventually be reclaimed by the GC.

Reply:
Ah :) Ok.  Thanks.

Reply:
Can't we use the ArrayBufferWriter here to reduce duplication between these methods?